### PR TITLE
Adds live migration procedures for sdn to ovnk

### PR DESCRIPTION
**Pull request by stevsmit**
This is a manual cherry pick of both f328ff3625a597ddaef66818fc9648e5cefeea39 and 0d14457b277ca3c4c90183b7be7e8838164063ea. This content of this PR does not exactly match its cherry picks. The content of this PR *does* match the content that currently exists in OCP docs version 4.16, as it should. I used these original PRs as a base and then copied the content from each module from version 4.16. 

For review: I suggest comparing this PR to the modules in 4.16. Changes suggested to this PR would need to be done in 4.16 and then cherry-picked back to 4.15, because the content must match. 

tl;dr: it is a direct copy of content that exists in version 4.16. 

Preview: https://81120--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.html

Version(s):
4.15

Issue:
Jira: https://issues.redhat.com/browse/OSDOCS-11834

Link to docs preview:

QE review:
- [ ] QE has approved this change.

### CODE REVIEW

**Comment by pliurh** [comment]
LGTM

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+The following procedure shows you how to check for egress IP resources, egress firewall resources, multicast enabled namespaces, and network policies when your deploying is using OVN-Kubernetes. If you had these resources on OpenShift SDN, you should check them after migration to ensure that they are working properly.
+
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have successfully migrated from OpenShift SDN to OVN-Kubernetes by using the limited live migration.
+
+.Procedure
+
+. As an {product-title} cluster administrator, check for egress firewall resources. You can do this by using the `oc` CLI, or by using the {product-title} web console.
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:18]: Colon leading into substeps, per ISG.
```suggestion
. As an {product-title} cluster administrator, check for egress firewall resources. You can do this by using the `oc` CLI, or by using the {product-title} web console:
```

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+.. To check for egress firewall resource by using the `oc` CLI tool:
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:20]: This and line 65 could be unordered, no? Since you choose either one (or both)?
```suggestion
* To check for egress firewall resource by using the `oc` CLI tool:
```

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    to:
+      cidrSelector: 192.168.0.0/16
+  - type: Deny
+    to:
+      cidrSelector: 0.0.0.0/0
+----
++
+Ensure that the behavior of this resource is intended because it could have changed after migration. For more information about egress firewalls, see "Configuring an egress firewall for a project".
+
+.. To check for egress firewall resources by using the {product-title} web console:
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:65]:
```suggestion
* To check for egress firewall resources by using the {product-title} web console:
```

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+... On the {product-title} web console, click *Observe* -> *Metrics*.
+
+... In the *Expression* box, type `ovnkube_controller_num_egress_firewall_rules` and click *Run queries*. If you have egress firewall resources, they are returned in the *Expression* box.
+
+. Check your cluster for egress IP resources. You can do this by using the `oc` CLI, or by using the {product-title} web console.
+
+.. To check for egress IPs by using the `oc` CLI tool:
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:73]: Same:
```suggestion
. Check your cluster for egress IP resources. You can do this by using the `oc` CLI, or by using the {product-title} web console:

* To check for egress IPs by using the `oc` CLI tool:
```

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  - 192.0.2.12
+  - 192.0.2.13
+  namespaceSelector:
+    matchLabels:
+      name: my-namespace
+----
++
+Repeat this for all egress IPs. Ensure that the behavior of each resource is intended because it could have changed after migration. For more information about EgressIPs, see "Configuring an EgressIP address".
+
+.. To check for egress IPs by using the {product-title} web console:
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:124]:
```suggestion
* To check for egress IPs by using the {product-title} web console:
```

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+... On the {product-title} web console, click *Observe* -> *Metrics*.
+
+... In the *Expression* box, type `ovnkube_clustermanager_num_egress_ips` and click *Run queries*. If you have egress firewall resources, they are returned in the *Expression* box.
+
+. Check your cluster for multicast enabled namespaces. You can only do this by using the `oc` CLI.
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:130]:
```suggestion
. Check your cluster for multicast enabled namespaces. You can only do this by using the `oc` CLI:
```
I won't keep harping on the "colon leading into substeps", but check for those throughout.

File: modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  Allowing egress traffic:
+    To Port: <any> (traffic allowed to all ports)
+    To:
+      IPBlock:
+        CIDR: 224.0.0.0/4
+        Except: 
+  Policy Types: Ingress, Egress
+----
++
+Ensure that the behavior of the network policy is as intended. Optimization for network policies differ between SDN and OVN-K, so users might need to adjust their policies to achieve optimal performance for different CNIs. For more information, see "About network policy".
```
Comment (adellape) [inline modules/checking-cluster-resources-after-initiating-limited-live-migration.adoc:223]: I do see a handful of other instances of "OVN-K" throughout the repo, but it makes me wonder. A [term glossary](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc) or SSG entry saying whether this shortname is permissible would be nice.

Also is "SDN" by itself OK here? Or should it be more specifically:
```suggestion
Ensure that the behavior of the network policy is as intended. Optimization for network policies differ between OpenShift SDN and OVN-K, so users might need to adjust their policies to achieve optimal performance for different CNIs. For more information, see "About network policy".
```

File: modules/nw-ovn-kubernetes-checking-live-migration-metrics.adoc [hunk 8]
```diff
@@ -0,0 +1,75 @@
+// Module included in the following assemblies:
+//
+// * networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="checking-live-migration-metrics"]
+= Checking limited live migration metrics 
+
+Metrics are available to monitor the progress of the limited live migration. Metrics can be viewed on the {product-title} web console, or by using the `oc` CLI.
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-checking-live-migration-metrics.adoc:9]: There are a lot of instances throughout the PR of "the `oc` CLI" that could instead be:
```suggestion
Metrics are available to monitor the progress of the limited live migration. Metrics can be viewed on the {product-title} web console, or by using the {oc-first}.
```

File: modules/nw-ovn-kubernetes-checking-live-migration-metrics.adoc [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+.Procedure
+
+. To view limited live migration metrics on the {product-title} web console:
+
+.. Click *Observe* -> *Metrics*.
+
+.. In the *Expression* box, type *openshift_network* and click the *openshift_network_operator_live_migration_procedure* option. 
+
+. To view metrics by using the `oc` CLI:
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-checking-live-migration-metrics.adoc:23]: Unordered for web console vs CLI?
```suggestion
* To view limited live migration metrics on the {product-title} web console:

.. Click *Observe* -> *Metrics*.

.. In the *Expression* box, type *openshift_network* and click the *openshift_network_operator_live_migration_procedure* option. 

* To view metrics by using the `oc` CLI:
```

File: modules/nw-ovn-kubernetes-live-migration-about.adoc [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+* Automatic node rebooting
+* Seamless transition from the OpenShift SDN network plugin to the OVN-Kubernetes network plugin
+
+Although a rollback procedure is provided, the limited live migration is intended to be a one-way process.
+
+include::snippets/sdn-deprecation-statement.adoc[]
+
+The following sections provide more information about the limited live migration method.
+
+[id="supported-platforms-live-migrating-ovn-kubernetes"]
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-live-migration-about.adoc:23]:
```suggestion
[id="supported-platforms-live-migrating-ovn-kubernetes_{context}"]
```

File: modules/nw-ovn-kubernetes-live-migration-about.adoc [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+* OpenShift SDN does not support IPsec. After the migration, cluster administrators can enable IPsec.
+
+* OpenShift SDN does not support IPv6. After the migration, cluster administrators can enable dual-stack.
+
+* The OpenShift SDN plugin allows application of the `NodeNetworkConfigurationPolicy` (NNCP) custom resource (CR) to the primary interface on a node. The OVN-Kubernetes network plugin does not have this capability.
+
+* The cluster MTU is the MTU value for pod interfaces. It is always less than your hardware MTU to account for the cluster network overlay overhead. The overhead is 100 bytes for OVN-Kubernetes and 50 bytes for OpenShift SDN.
++
+During the limited live migration, both OVN-Kubernetes and OpenShift SDN run in parallel. OVN-Kubernetes manages the cluster network of some nodes, while OpenShift SDN manages the cluster network of others. To ensure that cross-CNI traffic remains functional, the Cluster Network Operator updates the routable MTU to ensure that both CNIs share the same overlay MTU. As a result, after the migration has completed, the cluster MTU is 50 bytes less.
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-live-migration-about.adoc:67]:
```suggestion
During the limited live migration, both OVN-Kubernetes and OpenShift SDN run in parallel. OVN-Kubernetes manages the cluster network of some nodes, while OpenShift SDN manages the cluster network of others. To ensure that cross-CNI traffic remains functional, the Cluster Network Operator updates the routable MTU to ensure that both CNIs share the same overlay MTU. As a result, after the migration has completed, the cluster MTU is 50 bytes fewer.
```
?

File: modules/nw-ovn-kubernetes-live-migration-about.adoc [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+** `InternalTransitSwitchSubnet`
+** `internalJoinSubnet`
+
+* OVN-Kubernetes reserves the `100.64.0.0/16` and `100.88.0.0/16` IP address ranges. These subnets cannot be overlapped with any other internal or external network. If these IP addresses have been used by OpenShift SDN or any external networks that might communicate with this cluster, you must patch them to use a different IP address range before starting the limited live migration. See "Patching OVN-Kubernetes address ranges" for more information.
+
+* In most cases, the limited live migration is independent of the secondary interfaces of pods created by the Multus CNI plugin. However, if these secondary interfaces were set up on the default network interface controller (NIC) of the host, for example, using MACVLAN, IPVLAN, SR-IOV, or bridge interfaces with the default NIC as the control node, OVN-Kubernetes might encounter malfunctions. Users should remove such configurations before proceeding with the limited live migration.
+
+* When there are multiple NICs inside of the host, and the default route is not on the interface that has the Kubernetes NodeIP, you must use the offline migration instead.
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-live-migration-about.adoc:78]: `NodeIP` ?
node IP ?

File: modules/nw-ovn-kubernetes-migration-about.adoc [hunk 13]
```diff
@@ -3,38 +3,45 @@
...
+The following sections provide more information about the offline migration method.
 
-* Bare-metal hardware
-* {aws-first}
-* {gcp-first}
-* {ibm-cloud-name}
-* {azure-first}
-* {rh-openstack-first}
-* {vmw-first}
+[id="supported-platforms-offline-migrating-ovn-kubernetes"]
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-migration-about.adoc:22]:
```suggestion
[id="supported-platforms-offline-migrating-ovn-kubernetes_{context}"]
```

File: modules/nw-ovn-kubernetes-rollback-live.adoc [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+Cluster Network Operator (CNO):: Performs the following actions:
++
+--
+* Sets migration related fields in the `network.operator` custom resource (CR) and waits for routable MTUs to be applied to all nodes by the Machine Config Operator (MCO).
+* Patches the `network.operator` CR to set the migration mode to `Live` for OpenShiftSDN and deploys the OpenShiftSDN network plugin in migration mode.
+* Deploys OVN-Kubernetes with hybrid overlay enabled.
+* Waits for both CNI plugins to be deployed and updates the conditions in the status of the `network.config` CR.
+* Triggers the MCO to apply the new machine config to each machine config pool, which includes node cordoning, draining, and rebooting.
+* Removes migration-related fields from the `network.operator` CR and performs cleanup actions, such as deleting OpenShift SDN resources and redeploying OVN-Kubernetes in normal mode with the necessary configurations.
+* Waits for the OpenShiftSDN redeployment and updates the status conditions in the `network.config` CR to indicate migration completion.
```
Comment (adellape) [inline modules/nw-ovn-kubernetes-rollback-live.adoc:39]: Same question about `OpenShiftSDN` vs "OpenShift SDN"

File: modules/patching-ovnk-address-ranges.adoc [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+// * networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="patching-ovnk-address-ranges_{context}"]
+= Patching OVN-Kubernetes address ranges
+
+OVN-Kubernetes reserves the following IP address ranges:
+
+* `100.64.0.0/16`. This IP address range is used for the `internalJoinSubnet` parameter of OVN-Kubernetes by default. 
+* `100.88.0.0/16`. This IP address range is used for the `internalTransSwitchSubnet` parameter of OVN-Kubernetes by default.
```
Comment (adellape) [inline modules/patching-ovnk-address-ranges.adoc:12]: Nit
```suggestion
* `100.64.0.0/16`: This IP address range is used for the `internalJoinSubnet` parameter of OVN-Kubernetes by default. 
* `100.88.0.0/16`: This IP address range is used for the `internalTransSwitchSubnet` parameter of OVN-Kubernetes by default.
```

File: modules/patching-ovnk-address-ranges.adoc [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+If these IP addresses have been used by OpenShift SDN or any external networks that might communicate with this cluster, you must patch them to use a different IP address range before initiating the limited live migration.
+
+The following procedure can be used to patch CIDR ranges that are in use by OpenShift SDN if the migration was initially blocked.
+
+[NOTE]
+====
+This is an optional procedure and must only be used if the migration was blocked after using the `oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OVNKubernetes"}}'` command "Initiating the limited live migration process".
```
Comment (adellape) [inline modules/patching-ovnk-address-ranges.adoc:20]: Is there maybe a missing "from" before the section title reference? Also suggest moving the long command to its own block.
```suggestion
This is an optional procedure and must only be used if the migration was blocked after using the following command from "Initiating the limited live migration process":

[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' \
    --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OVNKubernetes"}}'` 
----
```

File: modules/patching-ovnk-address-ranges.adoc [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+====
+
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+
+.Procedure
+
+. If the `100.64.0.0/16` IP address range is already in use, enter the following command to patch it to a different range. The following example uses  `100.63.0.0/16`.
```
Comment (adellape) [inline modules/patching-ovnk-address-ranges.adoc:29]: White space and colon:
```suggestion
. If the `100.64.0.0/16` IP address range is already in use, enter the following command to patch it to a different range. The following example uses `100.63.0.0/16`.
```

File: modules/patching-ovnk-address-ranges.adoc [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
++
+[source,terminal]
+----
+$ oc patch network.operator.openshift.io cluster --type='merge' -p='{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"ipv4":{"internalJoinSubnet": "100.63.0.0/16"}}}}}'
+----
+
+. If the `100.88.0.0/16` IP address range is already in use, enter the following command to patch it to a different range. The following example uses  `100.99.0.0/16`.
```
Comment (adellape) [inline modules/patching-ovnk-address-ranges.adoc:36]:
```suggestion
. If the `100.88.0.0/16` IP address range is already in use, enter the following command to patch it to a different range. The following example uses `100.99.0.0/16`:
```

File: networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc [hunk 19]
```diff
@@ -6,7 +6,12 @@ include::_attributes/common-attributes.adoc[]
...
 
-As a cluster administrator, you can migrate to the OVN-Kubernetes network plugin from the OpenShift SDN network plugin.
+As a cluster administrator, you can migrate to the OVN-Kubernetes network plugin from the OpenShift SDN network plugin using the _offline_ migration method or the limited live migration method.
+
+[WARNING]
+====
+It is not possible to upgrade a cluster to {product-title} 4.17 if it is using the OpenShift SDN network plugin. You must migrate to the OVN-Kubernetes plugin before upgrading to {product-title} 4.17.
+====
 
 To learn more about OVN-Kubernetes, read xref:../../networking/ovn_kubernetes_network_provider/about-ovn-kubernetes#about-ovn-kubernetes[About the OVN-Kubernetes network plugin].
```
Comment (adellape) [inline networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc:16]:
```suggestion
To learn more about OVN-Kubernetes, see xref:../../networking/ovn_kubernetes_network_provider/about-ovn-kubernetes#about-ovn-kubernetes[About the OVN-Kubernetes network plugin].
```

**Comment by stevsmit** [comment]
@adellape I'll need to make these changes in 4.16 and CP back to 4.15 with a new PR
